examples/hidden_shift_algorithm.py

```diff
@@ -0,0 +1,140 @@
+import random
+import cirq
+"""Demonstrates a Hidden Shift algorithm.
+
+Let f and g be two function oracles {0,1}^N -> {0,1}^N  which are the same
+up to a hidden bit string s:
+
+g(x) = f(x ⨁ s)
+
+The Hidden Shift Algorithm determines s by quering the two oracles. The
+implementation in this example considers the following oracle:
+
+f(x) = Σ_i x_(2i-1) x_(2i)
+
+where x_i is the i-th bit of x.
+
+While a classical algorithm requires 2^(N/2) queries, the Hidden Shift
+Algorithm solves the problem in O(1) steps. We thus have an exponential
+reduction.
+
+=== REFERENCE ===
+[1] Wim van Dam, Sean Hallgreen, Lawrence Ip Quantum Algorithms for some
+Hidden Shift Problems. https://arxiv.org/abs/quant-ph/0211140
+[2] K Wrigth, et. a. Benchmarking an 11-qubit quantum computer.
+Nature Communications, 107(28):12446–12450, 2010. doi:10.1038/s41467-019-13534-2
+
+
+=== EXAMPLE OUTPUT ===
+Secret shift sequence: [1, 0, 0, 1, 0, 1]
+Circuit:
+(0, 0): ───H───X───@───X───H───@───H───M('result')───
+                   │           │       │
+(1, 0): ───H───────@───────H───@───H───M─────────────
+                                       │
+(2, 0): ───H───────@───────H───@───H───M─────────────
+                   │           │       │
+(3, 0): ───H───X───@───X───H───@───H───M─────────────
+                                       │
+(4, 0): ───H───────@───────H───@───H───M─────────────
+                   │           │       │
+(5, 0): ───H───X───@───X───H───@───H───M─────────────
+Sampled results:
+Counter({'100101': 100})
+Most common bitstring: 100101
+Found a match: True
+"""
+
+
+def set_qubits(qubit_count):
+    """Add the specified number of input qubits."""
+    input_qubits = [cirq.GridQubit(i, 0) for i in range(qubit_count)]
+    return input_qubits
+
+
+def make_oracle_f(qubits):
+    """Implement function {f(x) = Σ_i x_(2i-1) x_(2i)}."""
+    return [
+        cirq.CZ(qubits[2 * i], qubits[2 * i + 1])
```

Comment:
The function says `2i-1` but this cirq.CZ has `2i+1`. Why the discrepancy?

Reply:
Fixed indexing